Modules/socketmodule.c

```diff
@@ -3332,32 +3332,55 @@ sock_setsockopt(PyObject *self, PyObject *args)
 {
     PySocketSockObject *s = _PySocketSockObject_CAST(self);
 
+    Py_ssize_t arglen;
     int level;
     int optname;
     int res;
-    Py_buffer optval;
+    Py_buffer buffer;
     int flag;
     unsigned int optlen;
-    PyObject *none;
+    PyObject *optval;
+
+    arglen = PyTuple_Size(args);
+    switch (arglen) {
+        case 3:
+            if (!PyArg_ParseTuple(args, "iiO:setsockopt",
+                                  &level, &optname, &optval)) {
+                return NULL;
+            }
+            break;
+        case 4:
+            if (!PyArg_ParseTuple(args, "iiO!I:setsockopt",
+                                  &level, &optname, Py_TYPE(Py_None), &optval, &optlen)) {
+                return NULL;
+            }
+            break;
```

Comment:
We need `if (optval != Py_None) {error...}` here.

Reply:
Look at https://github.com/python/cpython/pull/107546/files#r2192102447.
The "O!" format along with Py_TYPE(Py_None) ensures that the type of the 3rd argument would be None.